File: backend/internal/models/binary.go
package models

import (
	"time"

	"github.com/google/uuid"
)

// BinaryType represents the type of binary (e.g., hashcat, john).
type BinaryType string

const (
	BinaryTypeHashcat BinaryType = "hashcat"
	BinaryTypeJohn    BinaryType = "john"
)

// CompressionType represents the archive compression type.
type CompressionType string

const (
	CompressionType7z    CompressionType = "7z"
	CompressionTypeZip   CompressionType = "zip"
	CompressionTypeTarGz CompressionType = "tar.gz"
	CompressionTypeTarXz CompressionType = "tar.xz"
)

// BinaryVersion mirrors the binary_versions table structure.
type BinaryVersion struct {
	ID                 int             `json:"id" db:"id"` // SERIAL PRIMARY KEY
	BinaryType         BinaryType      `json:"binary_type" db:"binary_type"`
	CompressionType    CompressionType `json:"compression_type" db:"compression_type"`
	SourceURL          string          `json:"source_url" db:"source_url"`
	FileName           string          `json:"file_name" db:"file_name"`
	MD5Hash            string          `json:"md5_hash" db:"md5_hash"`
	FileSize           int64           `json:"file_size" db:"file_size"` // BIGINT
	CreatedAt          time.Time       `json:"created_at" db:"created_at"`
	CreatedBy          uuid.UUID       `json:"created_by" db:"created_by"`
	IsActive           bool            `json:"is_active" db:"is_active"`
	LastVerifiedAt     time.Time       `json:"last_verified_at,omitempty" db:"last_verified_at"`       // Nullable
	VerificationStatus string          `json:"verification_status,omitempty" db:"verification_status"` // Nullable, VARCHAR(50)
}

// BinaryVersionBasic is a subset used for form data lists.
type BinaryVersionBasic struct {
	ID   int    `json:"id"`
	Name string `json:"name"` // Using FileName as Name for display
}

File: frontend/src/services/binary.ts
import { api } from './api';

export interface BinaryVersion {
  id: number;
  binary_type: 'hashcat' | 'john';
  compression_type: '7z' | 'zip' | 'tar.gz' | 'tar.xz';
  source_url: string;
  file_name: string;
  md5_hash: string;
  file_size: number;
  created_at: string;
  is_active: boolean;
  is_default: boolean;
  last_verified_at: string | null;
  verification_status: 'pending' | 'verified' | 'failed' | 'deleted';
}

export interface AddBinaryRequest {
  binary_type: 'hashcat' | 'john';
  compression_type: '7z' | 'zip' | 'tar.gz' | 'tar.xz';
  source_url: string;
  file_name: string;
  set_as_default?: boolean;
}

export const listBinaries = async () => {
  try {
    const response = await api.get<BinaryVersion[]>('/api/admin/binary');
    console.debug('Binary list response:', response);
    return response;
  } catch (error) {
    console.error('Error in listBinaries:', error);
    throw error;
  }
};

export const addBinary = (binary: AddBinaryRequest) => {
  return api.post<BinaryVersion>('/api/admin/binary', binary);
};

export const verifyBinary = (id: number) => {
  return api.post<void>(`/api/admin/binary/${id}/verify`);
};

export const deleteBinary = (id: number) => {
  return api.delete<void>(`/api/admin/binary/${id}`);
};

export const getBinary = (id: number) => {
  return api.get<BinaryVersion>(`/api/admin/binary/${id}`);
};

export const setDefaultBinary = (id: number) => {
  return api.put<{ message: string }>(`/api/admin/binary/${id}/set-default`);
}; 